src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AdvSimdHelper.cs

```diff
@@ -0,0 +1,128 @@
+
+            return mask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<sbyte> CreateEscapingMask_UnsafeRelaxedJavaScriptEncoder(Vector128<sbyte> sourceValue)
+        {
+            Debug.Assert(AdvSimd.IsSupported);
+
+            // Anything in the control characters range (except 0x7F), and anything above sbyte.MaxValue but less than or equal byte.MaxValue
+            // That's because anything between 128 and 255 (inclusive) will overflow and become negative.
+            Vector128<sbyte> mask = AdvSimd.CompareLessThan(sourceValue, s_spaceMaskSByte);
+
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_quotationMarkMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_reverseSolidusMaskSByte));
+
+            // Leftover control character in the ASCII range - 0x7F
+            // Since we are dealing with sbytes, 0x7F is the only value that would meet this comparison.
+            mask = AdvSimd.Or(mask, AdvSimd.CompareGreaterThan(sourceValue, s_tildeMaskSByte));
+
+            return mask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<sbyte> CreateEscapingMask_DefaultJavaScriptEncoderBasicLatin(Vector128<sbyte> sourceValue)
+        {
+            Debug.Assert(AdvSimd.IsSupported);
+
+            Vector128<sbyte> mask = CreateEscapingMask_UnsafeRelaxedJavaScriptEncoder(sourceValue);
+
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_ampersandMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_apostropheMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_plusSignMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_lessThanSignMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_greaterThanSignMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_graveAccentMaskSByte));
+
+            return mask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<short> CreateAsciiMask(Vector128<short> sourceValue)
+        {
+            Debug.Assert(AdvSimd.IsSupported);
+
+            // Anything above short.MaxValue but less than or equal char.MaxValue
+            // That's because anything between 32768 and 65535 (inclusive) will overflow and become negative.
+            Vector128<short> mask = AdvSimd.CompareLessThan(sourceValue, s_nullMaskInt16);
+
+            // Anything above the ASCII range
+            mask = AdvSimd.Or(mask, AdvSimd.CompareGreaterThan(sourceValue, s_maxAsciiCharacterMaskInt16));
+
+            return mask;
+        }
+
+        // Encodes an operation equivalent to Sse2.MoveMask
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int MoveMask(Vector128<byte> value)
```

Comment:
Please refer my comments [https://github.com/dotnet/runtime/pull/38653/files#r452562247](https://github.com/dotnet/runtime/pull/38653/files#r452562247) about using `static` variables.

Reply:
The particular method is being used primarily by the `NeedsEscaping` method which is being called in various places in the loop, so that would need to accept the bitmask parametrically as well. My only concern is whether the added complexity of passing around the bitmask could justify the hypothetical performance gains.

Note that other methods, such as `CreateEscapingMask_DefaultJavaScriptEncoderBasicLatin` in the same file is reading vectors from 10 static fields. We could lift those as well, but it just feels too error prone.

Reply:
You can try doing a quick experiment to see both the generated JIT code with and without static variables as well as measure performance to see if the complexity justifies the gain. But we don't hoist `static` variable access because we don't consider it as invariant inside a loop (https://github.com/dotnet/runtime/issues/35279#issuecomment-620125815) and hence that involves memory access.

Reply:
Perhaps you can still create those 2 variables static, but then save them inside a local variable outside the loop and pass them to the `AdvSimdMoveMask()` just to make sure that all call sites pass same value of `mostSignificantBitIsSet` and `s_bitMask128`.

Reply:
Would it be better to define masks as
```c#
       private static readonly sbyte s_ampersandMaskSByte = (sbyte)'&';
       private static readonly sbyte s_apostropheMaskSByte =(sbyte)'\'';
...
```
and then in `CreateEscapingMask_DefaultJavaScriptEncoderBasicLatin` have slightly more complex implementation
```c#
            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, Vector128.Create(s_ampersandMaskSByte)));
            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, Vector128.Create(s_apostropheMaskSByte)));
...
```

This would ensure that the JIT optimizes this correctly, since static readonly fields are treated as constants.

Reply:
`Vector128.Create` is already handled as a constant if all inputs are constant, so if we do `Vector128.Create((sbyte)'&')` it should generate fairly optimal code. It just may or may not hoist it for reuse if in a loop (so you may read the constant from memory repeatedly).